### ISSUE TITLE
Unable to embed ICC profile in JPEG

### ISSUE THREAD

**Issue by eroux** [haraldk/TwelveMonkeys#528]
The context is the following: I'm trying to achieve the following:
- read an image and its icc profile (not converting the image to the icc profile, just reading the two)
- write the image and its original icc profile (again, no conversion)

I first thought this would be a triviality but I've had no success so far. My current code is in [this file](https://github.com/buda-base/buda-iiif-server/blob/9e2c1f6af97ad23258fa528d4e508fc9a90d5eaf/src/test/java/io/bdrc/iiif/ImageTest.java), using [this image](https://github.com/buda-base/buda-iiif-server/blob/9e2c1f6af97ad23258fa528d4e508fc9a90d5eaf/src/test/resources/ORIGINAL_S3.jpg) which has an `Adobe RGB (1998)` color profile. The code can be summarized as follows:

```java
    public static void readAndWriteTwelveMonkeys(ImageInputStream iis) {
        ImageReader r = ...; // the twelvemonkeys one
        r.setInput(iis);
        
        IIOMetadata meta = r.getImageMetadata(0);
        // if I print the image metadata, I see that the color profile is correct (AdobeRGB)
        
        ImageTypeSpecifier its = r.getRawImageType(0);
        ImageReadParam p = r.getDefaultReadParam();
        p.setDestinationType(its);
        
        // color space is RGB (not sRGB), not sure if it's relevant
        System.out.println("is ColorSpace RGB? "+(its.getColorModel().getColorSpace().getType() == ColorSpace.TYPE_RGB));
                
        BufferedImage bi = r.read(0, p);

        Iterator<ImageWriter> itw2 = ImageIO.getImageWriters(new ImageTypeSpecifier(bi), "jpeg");
        ImageWriter iw2 = itw2.next();
        if (iw2 == null) {
            System.out.println("no writer for the image type of the buffered image");
        } else {
            ImageWriteParam wp = iw2.getDefaultWriteParam();
            wp.setDestinationType(its);
            ImageOutputStream out = ImageIO.createImageOutputStream(new File("test-regularjpg.jpg"));
            iw2.setOutput(out);
            iw2.write(null, new IIOImage(bi, null, meta), wp);
        }
        is.close();
        iis.close();
    }
```

which works but the output doesn't have the original icc profile embedded; even more strange it embeds an sRGB profile.

Is it generally possible to achieve what I want: with `ImageIO` in general (I'm starting to doubt it)? with TwelveMonkeys?

`java -version` says

```
openjdk version "11.0.5" 2019-10-15
```

**Edit:** I initially reported a silly configuration error, sorry

**Comment by eroux** [comment]
so, after a lot of debugging, it seems that the problem is in fact not really on the side of the writer which embeds icc profiles fine, the problem is that the TwelveMonkey reader performs a `ColorConvertOp` on the raster to convert it to `sRGB` instead of creating a `BufferedImage` with a `ColorModel` corresponding to the icc profile. Could there be an option to do that? 

Note: If anyone else stumbles upon that issue, I have worked around for it using another reader that doesn't convert the raster, see [this function](https://github.com/buda-base/buda-iiif-server/blob/3797d9be982ff222e74e372f6387a1ee219c859e/src/test/java/io/bdrc/iiif/ImageTest.java#L79).

**Comment by haraldk** [comment]
Hi Elie,

I believe what you are trying to do should be achievable using ImageIO, but unfortunately, these things are typically less than trivial...

~~First, I believe `r.getRawImageType(0)` will always return `null` for JPEG, as the "raw" type for JPEG is typically YCbCr and there is no Java `ColorSpace` instance matching this. So, I suggest removing this.  At least I get a `NullPointerException`  But perhaps your input is not a JPEG?~~ Hmm.. This behavior seems to be changed from Java 8 to Java 11. I need to dig a bit deeper to figure out if this *really* is the embedded CS though. I believe not, and even if it was, it would be redundant as it's already contained in the metadata, and you should still remove it,

Second, you check the color space *type*, both AdobeRGB1998 and sRGB are type RGB color spaces, so I don't think that's relevant. That's how it should be.

Finally, it seems that the ICC profile in the image (`BufferedImage`) takes precedence over the ICC profile in the metadata when writing. I believed that if you want to keep the pixels as is, and retain the color profile, you could have used `readRaster` instead of `read`, and pass the raster to the `IIOImage` constructor, along with the metadata containing the ICC profile. Unfortunately, this does not work right away either, as the ICC profile doesn't seem to be written, if writing raster... 

However, I managed to get the ICC profile written, by reading the raster, "manually" converting YCbCr to RGB, creating a color space from the ICC profile, then creating an image from that and write it.

Something like this (will work for YCbCr encoded JPEGs with ICC profile only! Modify to your needs). Unfortunately, this is way too much code for a task that shouldn't be this hard...

````java
ImageInputStream iis = ImageIO.createImageInputStream(input);
ImageReader ir = ImageIO.getImageReaders(iis).next();
ir.setInput(iis);

WritableRaster raster = (WritableRaster) ir.readRaster(0, null);

IIOMetadata metadata = ir.getImageMetadata(0);
IIOMetadataNode tree = (IIOMetadataNode) metadata.getAsTree(metadata.getNativeMetadataFormatName());

BufferedImage image = null;
NodeList iccs = tree.getElementsByTagName("app2ICC");

if (iccs.getLength() > 0) {
    IIOMetadataNode icc = (IIOMetadataNode) iccs.item(0);
    Object iccData = icc.getUserObject();

    if (iccData instanceof ICC_Profile) {
        ICC_Profile profile = (ICC_Profile) iccData;
        ColorModel cm = new ComponentColorModel(ColorSpaces.createColorSpace(profile), false, false, Transparency.OPAQUE, DataBuffer.TYPE_BYTE);

        byte[] data = ((DataBufferByte) raster.getDataBuffer()).getData();

        for (int i = 0; i < data.length; i += 3) {
            YCbCrConverter.convertYCbCr2RGB(data, data, i);
        }

        image = new BufferedImage(cm, raster, cm.isAlphaPremultiplied(), null);
    }
}

ImageWriter iw = ImageIO.getImageWritersByFormatName("jpeg").next();

try (ImageOutputStream out = ImageIO.createImageOutputStream(new File("test-icc-profile.jpg"))) {
    iw.setOutput(out);
    iw.write(null, new IIOImage(image, null, metadata), null);
}
````

Note that when reading the image back, the image will still be converted from AdobeRGB1998 to sRGB.

I think a useful new feature, would be to allow reading the image in the color space that it originally had. This should be doable, by exposing this color space from the `getImageTypes(..)` iterator.


Hope this helps,

-- 
Harald K

**Comment by eroux** [comment]
Thanks a lot, that's very helpful! I agree with you that reading the image in its original color space would be a huge help

**Comment by haraldk** [comment]
Seems I wrote my answer about the same time as your update, so I missed it. But I guess we reached the same conclusion, it would be best if we allow reading the image in the original color space as an option.

I'll see what I can do!


-- 
Harald K

**Comment by haraldk** [comment]
Hi,

I think I might have "accidentally" fixed this is #573 (TwelveMonkeys 3.7.0), as we now always return non-null from `getRawImageType`. The raw type will use the color space created from the ICC profile (given that it is compatible with the image data), and the same type is available in `getImageTypes` as well. 

If you find the time, please test the new functionality, and see if that works for you. If not, you may re-open the issue with your new findings. 😀 

Closing for now.


Best regards,

-- 
Harald K

**Comment by eroux** [comment]
it looks fixed indeed, thanks a lot!
